File: backend/src/main/java/com/urbanman/ecommerce/controller/AuthController.java
package com.urbanman.ecommerce.controller;

import com.urbanman.ecommerce.model.User;
import com.urbanman.ecommerce.repository.UserRepository;
import com.urbanman.ecommerce.config.JwtUtil;
import com.urbanman.ecommerce.service.EmailService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/auth")
@CrossOrigin(origins = {
        "https://e-commerce-1-ariz.onrender.com",
        "https://e-commerce-two-rouge-62.vercel.app",
        "https://e-commerce-1ttw.vercel.app"
})
public class AuthController {

    private final UserRepository userRepo;
    private final EmailService emailService;
    private final JwtUtil jwtUtil;
    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

    @Autowired
    public AuthController(UserRepository userRepo, EmailService emailService, JwtUtil jwtUtil) {
        this.userRepo = userRepo;
        this.emailService = emailService;
        this.jwtUtil = jwtUtil;
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody User user) {
        if (userRepo.findByEmail(user.getEmail()).isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Email already exists"));
        }
        user.setPassword(passwordEncoder.encode(user.getPassword()));
        User savedUser = userRepo.save(user);
        return ResponseEntity.ok(Map.of(
                "message", "User registered successfully",
                "user", Map.of(
                        "id", savedUser.getUserId(),
                        "name", savedUser.getName(),
                        "email", savedUser.getEmail()
                )
        ));
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> request) {
        String email = request.get("email");
        String password = request.get("password");

        if (email == null || password == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Email and password are required"));
        }

        return userRepo.findByEmail(email)
                .filter(u -> passwordEncoder.matches(password, u.getPassword()))
                .map(u -> {
                    String token = jwtUtil.generateToken(u.getEmail(), u.getRole().name());
                    return ResponseEntity.ok(Map.of(
                        "token", token,
                        "user", Map.of("id", u.getUserId(), "name", u.getName(), "email", u.getEmail(), "role", u.getRole().name())
                    ));
                })
                .orElse(ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("error", "Invalid credentials")));
    }

    @PostMapping("/forgot-password")
    public ResponseEntity<?> forgotPassword(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        if (email == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Email is required"));
        }
        return userRepo.findByEmail(email).map(user -> {
            String tempPass = "Reset@" + (int)(1000 + Math.random() * 9000);
            user.setPassword(passwordEncoder.encode(tempPass));
            userRepo.save(user);
            emailService.sendPasswordReset(email, user.getName() != null ? user.getName() : "Customer", tempPass);
            return ResponseEntity.ok(Map.of("message", "Password reset email sent"));
        }).orElse(ResponseEntity.badRequest().body(Map.of("error", "No account found with this email")));
    }
}

File: backend/controllers/authController.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const User = require('../models/User');
require('dotenv').config();

const SECRET = process.env.JWT_SECRET || 'karunada-ecommerce-secret-key-2026-secure-jwt-token-signing-key-minimum-512-bits';
const EXPIRATION = Number(process.env.JWT_EXPIRATION || 86400000);

const register = async (req, res) => {
  try {
    const { name, email, phone, password, role } = req.body;
    
    if (!email || !password) {
      return res.status(400).json({ error: 'Email and password are required' });
    }

    const existingUser = await User.findOne({ where: { email } });
    if (existingUser) {
      return res.status(400).json({ error: 'Email already exists' });
    }

    const hashedPassword = await bcrypt.hash(password, 10);
    const user = await User.create({
      name,
      email,
      phone,
      password: hashedPassword,
      role: role || 'USER'
    });

    return res.status(200).json({
      message: 'User registered successfully',
      user: {
        id: user.userId,
        name: user.name,
        email: user.email
      }
    });
  } catch (error) {
    console.error('Register error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

const login = async (req, res) => {
  try {
    const { email, password } = req.body;

    if (!email || !password) {
      return res.status(400).json({ error: 'Email and password are required' });
    }

    const user = await User.findOne({ where: { email } });
    if (!user) {
      return res.status(401).json({ error: 'Invalid credentials' });
    }

    const isMatch = await bcrypt.compare(password, user.password);
    if (!isMatch) {
      return res.status(401).json({ error: 'Invalid credentials' });
    }

    if (!user.active) {
      return res.status(403).json({ error: 'Account is deactivated' });
    }

    const token = jwt.sign(
      { id: user.userId, email: user.email, role: user.role },
      SECRET,
      { expiresIn: EXPIRATION / 1000 }
    );

    return res.status(200).json({
      token,
      user: {
        id: user.userId,
        name: user.name,
        email: user.email,
        role: user.role
      }
    });
  } catch (error) {
    console.error('Login error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

const forgotPassword = async (req, res) => {
  try {
    const { email } = req.body;
    if (!email) {
      return res.status(400).json({ error: 'Email is required' });
    }

    const user = await User.findOne({ where: { email } });
    if (!user) {
      return res.status(400).json({ error: 'No account found with this email' });
    }

    const tempPass = 'Reset@' + Math.floor(1000 + Math.random() * 9000);
    const hashedTempPassword = await bcrypt.hash(tempPass, 10);
    
    user.password = hashedTempPassword;
    await user.save();

    // Mocking email log as done in Spring Boot
    console.log(`[MOCK EMAIL] Sending password reset code to ${email} (${user.name}): Temp password is ${tempPass}`);

    return res.status(200).json({ message: 'Password reset email sent (Mocked)' });
  } catch (error) {
    console.error('Forgot password error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

module.exports = {
  register,
  login,
  forgotPassword
};
